vertx-web-client/src/main/java/io/vertx/ext/web/client/impl/cache/CacheControl.java

```diff
@@ -0,0 +1,176 @@
+/*
+ * Copyright 2021 Red Hat, Inc.
+ *
+ *  All rights reserved. This program and the accompanying materials
+ *  are made available under the terms of the Eclipse Public License v1.0
+ *  and Apache License v2.0 which accompanies this distribution.
+ *
+ *  The Eclipse Public License is available at
+ *  http://www.eclipse.org/legal/epl-v10.html
+ *
+ *  The Apache License v2.0 is available at
+ *  http://www.opensource.org/licenses/apache2.0.php
+ *
+ *  You may elect to redistribute this code under either of these licenses.
+ */
+package io.vertx.ext.web.client.impl.cache;
+
+import io.netty.handler.codec.DateFormatter;
+import io.vertx.core.MultiMap;
+import io.vertx.core.http.HttpHeaders;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+/**
+ * Parse HTTP headers to determine if and how to cache a response.
+ *
+ * @author <a href="mailto:craigday3@gmail.com">Craig Day</a>
+ */
+public class CacheControl {
+
+  private final Set<CacheControlDirective> directives;
+  private final Map<CacheControlDirective, Long> timeDirectives;
+  private final Instant expires;
+  private final Instant date;
+  private final String etag;
+  private final String vary;
+  private final long maxAge;
+
+  static CacheControl parse(MultiMap headers) {
+    return new CacheControl(headers);
+  }
+
+  private CacheControl(MultiMap headers) {
+    this.directives = new HashSet<>();
+    this.timeDirectives = new HashMap<>();
+    this.etag = headers.get(HttpHeaders.ETAG);
+    this.vary = headers.get(HttpHeaders.VARY);
+
+    if (headers.contains(HttpHeaders.DATE)) {
+      this.date = DateFormatter.parseHttpDate(headers.get(HttpHeaders.DATE)).toInstant();
+    } else {
+      this.date = Instant.now();
+    }
+
+    if (headers.contains(HttpHeaders.EXPIRES)) {
+      this.expires = DateFormatter.parseHttpDate(headers.get(HttpHeaders.EXPIRES)).toInstant();
+    } else {
+      this.expires = null;
+    }
+
+    // Order of operations matters here. We do plain assignment above, but then we must parse
+    // the Cache-Control header before we can compute max age.
+    parseAllCacheControl(headers);
+    this.maxAge = computeMaxAge();
+  }
+
+  public Set<CacheControlDirective> getDirectives() {
+    return directives;
+  }
+
+  public Map<CacheControlDirective, Long> getTimeDirectives() {
+    return timeDirectives;
+  }
+
+  public String getEtag() {
+    return etag;
+  }
+
+  public long getMaxAge() {
+    return maxAge;
+  }
+
+  public Set<CharSequence> variations() {
+    if (vary == null) {
+      return Collections.emptySet();
+    } else {
+      Set<CharSequence> variations = new HashSet<>();
+      for (String variation : vary.split(",")) {
```

Comment:
we should avoid using `split()` for performance reason, instead we should split this using string indexOf(char)

Reply:
Updated, let me know if you're happier with the new splitting.